rapid7_insight_agent/icon_rapid7_insight_agent/util/graphql_api/api_connection.py

```diff
@@ -0,0 +1,234 @@
+            "variables": {
+                "orgId": self.org_key,
+                "nextCursor": next_cursor
+            }
+        }
+        results_object = self._post_payload(payload)
+
+        has_next_page = results_object.get("data").get("organization").get("assets").get("pageInfo").get(
+            "hasNextPage")
+
+        next_agents = self._get_agents_from_result_object(results_object)
+
+        agents.extend(next_agents)
+
+        return has_next_page, results_object
+
+    def _find_agent_in_agents(self, agents, agent_input, agent_type):
+        """
+        Given a list of agent objects, find the agent that matches our input
+
+        :param agents: list (agents)
+        :param agent_input: String (Input value to look for)
+        :param agent_type: String (What type of input to look for, MAC, IP_ADDRESS, or HOSTNAME)
+
+        :return: dict (agent object)
+        """
+        self.logger.info(f"Searching for: {agent_input}")
+        self.logger.info(f"Search type: {agent_type}")
+        for agent in agents:
+            if agent_type == agent_typer.IP_ADDRESS:
+                if agent_input == agent.get("primaryAddress").get("ip"):
+                    return agent
+            elif agent_type == agent_typer.HOSTNAME:
+                # In this case, we have an alpha/numeric value. This could be the ID or the Hostname. Need to check both
+                if agent_input == agent.get("id"):
+                    return agent
+                for host_name in agent.get("hostNames"):
+                    if agent_input == host_name.get("name"):
+                        return agent
+            elif agent_type == agent_typer.MAC_ADDRESS:
+                # Mac addresses can come in with : or - as a separator, remove all of it and compare
+                stripped_input_mac = agent_input.replace("-", "").replace(":", "")
+                stripped_target_mac = agent.get("primaryAddress").get("mac").replace("-", "").replace(":", "")
+                if stripped_input_mac == stripped_target_mac:
+                    return agent
+            else:
+                raise APIException(cause="Could not determine agent type.",
+                                   assistance=f"Agent {agent_input} was not a Mac, IP, or Hostname.")
+
+        raise APIException(cause=f"Could not find agent matching {agent_input} of type {agent_type}.",
+                           assistance=f"Check the agent input value and try again.")
+
+    def _get_agents_from_result_object(self, results_object):
+        """
+        This will extract an agent object from the objec that's returned from the API
+
+        :param results_object: dict (API result payload)
+        :return: dict (agent object)
+        """
+        agent_list = []
+
+        edges = results_object.get("data").get("organization").get("assets").get("edges")
```

Comment:
This may be a better spot to access via key and then try/except on `KeyError` -  if any of these keys don't exist then you will get a `TypeError: 'NoneType' object is not subscriptable` which won't look nice